varlink/src/server.rs

```diff
@@ -194,13 +201,21 @@ impl Listener {
         }
 
         match self {
-            &Listener::TCP(Some(ref l), _) => {
-                let (s, _addr) = l.accept().map_err(map_context!())?;
-                Ok(Box::new(s))
+            &Listener::TCP(Some(ref l), accepted) => {
+                if accepted {
+                    unsafe { Ok(Box::new(TcpStream::from_raw_fd(l.as_raw_fd()))) }
```

Comment:
Without digging into this code a lot more, there's a reason those functions are unsafe and that reason is exactly what looks like a bug here: The previous listener still thinks it owns the file descriptor.

We can avoid this by deconstructing into an `OwnedFd` like `Box::new(TcpStream::from(l.into::<OwnedFd>()))` or so.

But I think we need to back up on this, I'll comment on the original ticket.